providers/implementations/signature/ecdsa_sig.c

```diff
@@ -222,105 +342,137 @@ static int ecdsa_sign(void *vctx, unsigned char *sig, size_t *siglen,
     return 1;
 }
 
-static int ecdsa_verify(void *vctx, const unsigned char *sig, size_t siglen,
-                        const unsigned char *tbs, size_t tbslen)
+static int ecdsa_signverify_message_update(void *vctx,
+                                         const unsigned char *data,
+                                         size_t datalen)
 {
     PROV_ECDSA_CTX *ctx = (PROV_ECDSA_CTX *)vctx;
 
-    if (!ossl_prov_is_running() || (ctx->mdsize != 0 && tbslen != ctx->mdsize))
+    if (ctx == NULL)
         return 0;
 
-    return ECDSA_verify(0, tbs, tbslen, sig, siglen, ctx->ec);
+    return EVP_DigestUpdate(ctx->mdctx, data, datalen);
 }
 
-static int ecdsa_setup_md(PROV_ECDSA_CTX *ctx, const char *mdname,
-                          const char *mdprops, const char *desc)
+static int ecdsa_sign_message_final(void *vctx, unsigned char *sig,
+                                  size_t *siglen, size_t sigsize)
 {
-    EVP_MD *md = NULL;
-    size_t mdname_len;
-    int md_nid, md_size;
-    WPACKET pkt;
-
-    if (mdname == NULL)
-        return 1;
+    PROV_ECDSA_CTX *ctx = (PROV_ECDSA_CTX *)vctx;
+    unsigned char digest[EVP_MAX_MD_SIZE];
+    unsigned int dlen = 0;
 
-    mdname_len = strlen(mdname);
-    if (mdname_len >= sizeof(ctx->mdname)) {
-        ERR_raise_data(ERR_LIB_PROV, PROV_R_INVALID_DIGEST,
-                       "%s exceeds name buffer length", mdname);
+    if (!ossl_prov_is_running() || ctx == NULL)
         return 0;
-    }
-    if (mdprops == NULL)
-        mdprops = ctx->propq;
-    md = EVP_MD_fetch(ctx->libctx, mdname, mdprops);
-    if (md == NULL) {
-        ERR_raise_data(ERR_LIB_PROV, PROV_R_INVALID_DIGEST,
-                       "%s could not be fetched", mdname);
+    if (ctx->mdctx == NULL)
         return 0;
-    }
-    md_size = EVP_MD_get_size(md);
-    if (md_size <= 0) {
-        ERR_raise_data(ERR_LIB_PROV, PROV_R_INVALID_DIGEST,
-                       "%s has invalid md size %d", mdname, md_size);
-        goto err;
-    }
-    md_nid = ossl_digest_get_approved_nid(md);
-    if (md_nid < 0) {
-        ERR_raise_data(ERR_LIB_PROV, PROV_R_DIGEST_NOT_ALLOWED,
-                       "digest=%s", mdname);
-        goto err;
-    }
-    /* XOF digests don't work */
-    if ((EVP_MD_get_flags(md) & EVP_MD_FLAG_XOF) != 0) {
-        ERR_raise(ERR_LIB_PROV, PROV_R_XOF_DIGESTS_NOT_ALLOWED);
-        goto err;
-    }
+    /*
+     * If sig is NULL then we're just finding out the sig size. Other fields
+     * are ignored. Defer to ecdsa_sign.
+     */
+    if (sig != NULL
+        && !EVP_DigestFinal_ex(ctx->mdctx, digest, &dlen))
+        return 0;
+    ctx->flag_allow_md = 1;
```

Comment:
Should this be set back to 1 if sigalg is 1?

Reply:
Nope.  I'll drop this line, which is safe since it's also in `ecdsa_digest_sign_final()` and `ecdsa_digest_verify_final()`